File: admin_app/admin_flutter/lib/models/production_order.dart
class ProductionOrder {
  const ProductionOrder({
    required this.id,
    required this.productId,
    required this.productName,
    required this.status,
    required this.quantity,
    required this.producedQuantity,
    required this.unit,
    required this.producedAt,
    this.number,
    this.userId,
    this.completedByUserId,
    this.canceledByUserId,
    this.unitCost,
    this.totalCost,
    this.estimatedUnitCost,
    this.estimatedTotalCost,
    this.dueDate,
    this.notes,
    this.cancellationReason,
    this.startedAt,
    this.completedAt,
    this.canceledAt,
    this.components = const [],
  });

  final int id;
  final String? number;
  final int productId;
  final String productName;
  final int? userId;
  final int? completedByUserId;
  final int? canceledByUserId;
  final String status;
  final double quantity;
  final double producedQuantity;
  final String unit;
  final double? unitCost;
  final double? totalCost;
  final double? estimatedUnitCost;
  final double? estimatedTotalCost;
  final String? dueDate;
  final String? notes;
  final String? cancellationReason;
  final DateTime? startedAt;
  final DateTime? completedAt;
  final DateTime? canceledAt;
  final DateTime producedAt;
  final List<ProductionOrderComponent> components;

  factory ProductionOrder.fromJson(Map<String, dynamic> json) {
    return ProductionOrder(
      id: json['id'] as int,
      number: json['number'] as String?,
      productId: json['product_id'] as int,
      productName: json['product_name'] as String,
      userId: json['user_id'] as int?,
      completedByUserId: json['completed_by_user_id'] as int?,
      canceledByUserId: json['canceled_by_user_id'] as int?,
      status: json['status'] as String? ?? 'concluida',
      quantity: _toDouble(json['quantity']),
      producedQuantity: _toDouble(json['produced_quantity']),
      unit: json['unit'] as String? ?? 'un',
      unitCost: _toNullableDouble(json['unit_cost']),
      totalCost: _toNullableDouble(json['total_cost']),
      estimatedUnitCost: _toNullableDouble(json['estimated_unit_cost']),
      estimatedTotalCost: _toNullableDouble(json['estimated_total_cost']),
      dueDate: json['due_date'] as String?,
      notes: json['notes'] as String?,
      cancellationReason: json['cancellation_reason'] as String?,
      startedAt: _toDateTime(json['started_at']),
      completedAt: _toDateTime(json['completed_at']),
      canceledAt: _toDateTime(json['canceled_at']),
      producedAt: DateTime.parse(json['produced_at'] as String).toLocal(),
      components: (json['components'] as List<dynamic>? ?? [])
          .map(
            (item) =>
                ProductionOrderComponent.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
    );
  }
}

class ProductionOrderComponent {
  const ProductionOrderComponent({
    required this.id,
    required this.componentProductId,
    required this.componentName,
    required this.quantity,
    required this.unit,
    required this.wastePercent,
    this.unitCost,
    this.totalCost,
  });

  final int id;
  final int componentProductId;
  final String componentName;
  final double quantity;
  final String unit;
  final double wastePercent;
  final double? unitCost;
  final double? totalCost;

  factory ProductionOrderComponent.fromJson(Map<String, dynamic> json) {
    return ProductionOrderComponent(
      id: json['id'] as int,
      componentProductId: json['component_product_id'] as int,
      componentName: json['component_name'] as String,
      quantity: _toDouble(json['quantity']),
      unit: json['unit'] as String? ?? 'un',
      wastePercent: _toDouble(json['waste_percent']),
      unitCost: _toNullableDouble(json['unit_cost']),
      totalCost: _toNullableDouble(json['total_cost']),
    );
  }
}

class ProductionOrderPayload {
  const ProductionOrderPayload({
    required this.productId,
    required this.quantity,
    this.dueDate,
    this.notes,
  });

  final int productId;
  final double quantity;
  final String? dueDate;
  final String? notes;

  Map<String, dynamic> toJson() {
    return {
      'product_id': productId,
      'quantity': quantity,
      'due_date': _emptyToNull(dueDate),
      'notes': _emptyToNull(notes),
    };
  }
}

class ProductionOrderPreview {
  const ProductionOrderPreview({
    required this.productId,
    required this.quantity,
    required this.unit,
    required this.components,
    this.estimatedUnitCost,
    this.estimatedTotalCost,
  });

  final int productId;
  final double quantity;
  final String unit;
  final double? estimatedUnitCost;
  final double? estimatedTotalCost;
  final List<ProductionOrderComponentPreview> components;

  factory ProductionOrderPreview.fromJson(Map<String, dynamic> json) {
    return ProductionOrderPreview(
      productId: json['product_id'] as int,
      quantity: _toDouble(json['quantity']),
      unit: json['unit'] as String? ?? 'un',
      estimatedUnitCost: _toNullableDouble(json['estimated_unit_cost']),
      estimatedTotalCost: _toNullableDouble(json['estimated_total_cost']),
      components: (json['components'] as List<dynamic>? ?? [])
          .map(
            (item) => ProductionOrderComponentPreview.fromJson(
              item as Map<String, dynamic>,
            ),
          )
          .toList(),
    );
  }
}

class ProductionOrderComponentPreview {
  const ProductionOrderComponentPreview({
    required this.componentProductId,
    required this.componentName,
    required this.requestedQuantity,
    required this.requestedUnit,
    required this.requiredQuantity,
    required this.stockQuantity,
    required this.unit,
    required this.wastePercent,
    required this.enoughStock,
    this.unitCost,
    this.totalCost,
  });

  final int componentProductId;
  final String componentName;
  final double requestedQuantity;
  final String requestedUnit;
  final double requiredQuantity;
  final double stockQuantity;
  final String unit;
  final double wastePercent;
  final double? unitCost;
  final double? totalCost;
  final bool enoughStock;

  factory ProductionOrderComponentPreview.fromJson(Map<String, dynamic> json) {
    return ProductionOrderComponentPreview(
      componentProductId: json['component_product_id'] as int,
      componentName: json['component_name'] as String,
      requestedQuantity: _toDouble(json['requested_quantity']),
      requestedUnit: json['requested_unit'] as String? ?? 'un',
      requiredQuantity: _toDouble(json['required_quantity']),
      stockQuantity: _toDouble(json['stock_quantity']),
      unit: json['unit'] as String? ?? 'un',
      wastePercent: _toDouble(json['waste_percent']),
      unitCost: _toNullableDouble(json['unit_cost']),
      totalCost: _toNullableDouble(json['total_cost']),
      enoughStock: json['enough_stock'] as bool? ?? false,
    );
  }
}

class ProductionOrderCompletePayload {
  const ProductionOrderCompletePayload({this.producedQuantity, this.notes});

  final double? producedQuantity;
  final String? notes;

  Map<String, dynamic> toJson() {
    return {
      'produced_quantity': producedQuantity,
      'notes': _emptyToNull(notes),
    };
  }
}

double _toDouble(Object? value) => _toNullableDouble(value) ?? 0;

double? _toNullableDouble(Object? value) {
  if (value == null) return null;
  if (value is num) return value.toDouble();
  return double.tryParse(value.toString());
}

DateTime? _toDateTime(Object? value) {
  if (value == null) return null;
  return DateTime.tryParse(value.toString())?.toLocal();
}

String? _emptyToNull(String? value) {
  final trimmed = value?.trim();
  if (trimmed == null || trimmed.isEmpty) return null;
  return trimmed;
}

File: backend/app/models/production_order.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import Date, DateTime, ForeignKey, Numeric, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class ProductionOrder(Base):
    __tablename__ = "production_orders"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    number: Mapped[Optional[str]] = mapped_column(String(40), unique=True, index=True)
    product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="RESTRICT"), index=True)
    user_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    completed_by_user_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    canceled_by_user_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="planejada")
    quantity: Mapped[Decimal] = mapped_column(Numeric(12, 3), nullable=False)
    produced_quantity: Mapped[Decimal] = mapped_column(Numeric(12, 3), nullable=False, default=0)
    unit: Mapped[str] = mapped_column(String(20), nullable=False, default="un")
    unit_cost: Mapped[Optional[Decimal]] = mapped_column(Numeric(12, 2))
    total_cost: Mapped[Optional[Decimal]] = mapped_column(Numeric(12, 2))
    estimated_unit_cost: Mapped[Optional[Decimal]] = mapped_column(Numeric(12, 2))
    estimated_total_cost: Mapped[Optional[Decimal]] = mapped_column(Numeric(12, 2))
    due_date: Mapped[Optional[date]] = mapped_column(Date)
    notes: Mapped[Optional[str]] = mapped_column(Text)
    cancellation_reason: Mapped[Optional[str]] = mapped_column(Text)
    started_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    completed_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    canceled_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    produced_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        index=True,
    )

    product = relationship("Product")
    user = relationship("User", foreign_keys=[user_id])
    completed_by_user = relationship("User", foreign_keys=[completed_by_user_id])
    canceled_by_user = relationship("User", foreign_keys=[canceled_by_user_id])
    components = relationship(
        "ProductionOrderComponent",
        back_populates="production_order",
        cascade="all, delete-orphan",
        lazy="selectin",
    )


class ProductionOrderComponent(Base):
    __tablename__ = "production_order_components"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    production_order_id: Mapped[int] = mapped_column(
        ForeignKey("production_orders.id", ondelete="CASCADE"),
        nullable=False,
    )
    component_product_id: Mapped[int] = mapped_column(ForeignKey("products.id", ondelete="RESTRICT"))
    component_name: Mapped[str] = mapped_column(String(180), nullable=False)
    quantity: Mapped[Decimal] = mapped_column(Numeric(12, 3), nullable=False)
    unit: Mapped[str] = mapped_column(String(20), nullable=False, default="un")
    waste_percent: Mapped[Decimal] = mapped_column(Numeric(7, 2), nullable=False, default=0)
    unit_cost: Mapped[Optional[Decimal]] = mapped_column(Numeric(12, 2))
    total_cost: Mapped[Optional[Decimal]] = mapped_column(Numeric(12, 2))

    production_order = relationship("ProductionOrder", back_populates="components")
    component_product = relationship("Product")
